/l-amlogic/system/core/debuggerd$ git diff
diff --git a/tombstone.cpp b/tombstone.cpp
index 0c1b80f..3403cb2 100644
--- a/tombstone.cpp
+++ b/tombstone.cpp
@@ -317,9 +317,10 @@ static void dump_stack(Backtrace* backtrace, log_t* log) {
       size_t words = frame->stack_size / sizeof(word_t);
       if (words == 0) {
         words = 1;
-      } else if (words > STACK_WORDS) {
-        words = STACK_WORDS;
       }
+      /*else if (words > STACK_WORDS) {
+        words = STACK_WORDS;
+      }*/
       dump_stack_segment(backtrace, log, &sp, words, i);
     }
   }
